src/number_theory/padics/padic_norm.lean

```diff
@@ -52,131 +52,153 @@ open_locale rat
 open multiplicity
 
 /--
-For `p ≠ 1`, the p-adic valuation of an integer `z ≠ 0` is the largest natural number `n` such that
-p^n divides z.
+For `p ≠ 1`, the p-adic valuation of a natural `n ≠ 0` is the largest natural number `k` such that
+p^k divides z.
 
-`padic_val_rat` defines the valuation of a rational `q` to be the valuation of `q.num` minus the
-valuation of `q.denom`.
-If `q = 0` or `p = 1`, then `padic_val_rat p q` defaults to 0.
+If `n = 0` or `p = 1`, then `padic_val_nat p q` defaults to 0.
 -/
-def padic_val_rat (p : ℕ) (q : ℚ) : ℤ :=
-if h : q ≠ 0 ∧ p ≠ 1
-then (multiplicity (p : ℤ) q.num).get
-    (multiplicity.finite_int_iff.2 ⟨h.2, rat.num_ne_zero_of_ne_zero h.1⟩) -
-  (multiplicity (p : ℤ) q.denom).get
-    (multiplicity.finite_int_iff.2 ⟨h.2, by exact_mod_cast rat.denom_ne_zero _⟩)
+def padic_val_nat (p : ℕ) (n : ℕ) : ℕ :=
+if h : p ≠ 1 ∧ 0 < n
+then (multiplicity p n).get (multiplicity.finite_nat_iff.2 h)
 else 0
 
-/--
-A simplification of the definition of `padic_val_rat p q` when `q ≠ 0` and `p` is prime.
--/
-lemma padic_val_rat_def (p : ℕ) [hp : fact p.prime] {q : ℚ} (hq : q ≠ 0) : padic_val_rat p q =
-  (multiplicity (p : ℤ) q.num).get (finite_int_iff.2 ⟨hp.1.ne_one, rat.num_ne_zero_of_ne_zero hq⟩) -
-  (multiplicity (p : ℤ) q.denom).get
-    (finite_int_iff.2 ⟨hp.1.ne_one, by exact_mod_cast rat.denom_ne_zero _⟩) :=
-dif_pos ⟨hq, hp.1.ne_one⟩
-
-namespace padic_val_rat
+namespace padic_val_nat
 open multiplicity
 variables {p : ℕ}
 
-/--
-`padic_val_rat p q` is symmetric in `q`.
--/
-@[simp] protected lemma neg (q : ℚ) : padic_val_rat p (-q) = padic_val_rat p q :=
+/-- `padic_val_nat p 0` is 0 for any `p`. -/
+@[simp] protected lemma zero : padic_val_nat p 0 = 0 :=
+by simp [padic_val_nat]
+
+/-- `padic_val_nat p 1` is 0 for any `p`. -/
+@[simp] protected lemma one : padic_val_nat p 1 = 0 :=
+by unfold padic_val_nat; split_ifs; simp *
+
+/-- For `p ≠ 0, p ≠ 1, `padic_val_rat p p` is 1. -/
+@[simp] lemma self (hp : 1 < p) : padic_val_nat p p = 1 :=
 begin
-  unfold padic_val_rat,
-  split_ifs,
-  { simp [-add_comm]; refl },
-  { exfalso, simp * at * },
-  { exfalso, simp * at * },
-  { refl }
+  have neq_one : (¬ p = 1) ↔ true,
+  { exact iff_of_true ((ne_of_lt hp).symm) trivial, },
+  have eq_zero_false : (p = 0) ↔ false,
+  { exact iff_false_intro ((ne_of_lt (trans zero_lt_one hp)).symm) },
+  simp [padic_val_nat, neq_one, eq_zero_false],
 end
 
+end padic_val_nat
+
 /--
-`padic_val_rat p 0` is 0 for any `p`.
+For `p ≠ 1`, the p-adic valuation of an integer `z ≠ 0` is the largest natural number `k` such that
+p^k divides z.
+
+If `x = 0` or `p = 1`, then `padic_val_int p q` defaults to 0.
 -/
+def padic_val_int (p : ℕ) (z : ℤ) : ℕ :=
+padic_val_nat p (z.nat_abs)
```

Comment:
I would like to have a lemma that this def is equivalent to `(multiplicity (p : ℤ) z).get _` under suitable assumptions.

Reply:
Ok, I added `padic_val_int.defn` to do this